### ISSUE TITLE
aleph.http.request could do a better job erroring if when mandatory arguments are missing

### ISSUE THREAD

**Issue by lvh** [clj-commons/aleph#162]
Turns out I forgot to pass `:method`; got this:

```
2015-Apr-14 10:37:55 -0400 zygalski.local DEBUG [icecap.handlers.http] - (request args) << … >>
Apr 14, 2015 10:37:55 AM manifold.stream invoke
SEVERE: error in consume
java.lang.NullPointerException
    at clojure.core$name.invoke(core.clj:1522)
    at aleph.http.core$ring_request__GT_netty_request.invoke(core.clj:157)
    at aleph.http.client$http_connection$fn__18161$fn__18162.invoke(client.clj:278)
    at manifold.stream.Callback.put(stream.clj:431)
    at manifold.stream.graph$async_send.invoke(graph.clj:50)
    at manifold.stream.graph$async_connect$this__13640.invoke(graph.clj:172)
    at manifold.stream.graph$async_connect$this__13640$fn__13641$fn__13642.invoke(graph.clj:153)
    at clojure.core$trampoline.invoke(core.clj:6033)
    at manifold.stream.graph$async_connect$this__13640$fn__13641.invoke(graph.clj:153)
    at manifold.deferred.Listener$f__12755__auto____12960.invoke(deferred.clj:158)
    at manifold.deferred.Listener.onSuccess(deferred.clj:158)
    at manifold.deferred.Deferred$fn__13037.invoke(deferred.clj:324)
    at manifold.deferred.Deferred.success(deferred.clj:324)
    at manifold.deferred$success_BANG_.invoke(deferred.clj:184)
    at manifold.stream.default.Stream$fn__13733.invoke(default.clj:96)
    at manifold.stream.default.Stream.put(default.clj:95)
    at manifold.stream.default.Stream.put(default.clj:119)
    at aleph.http.client$http_connection$fn__18161$fn__18173$fn__18174.invoke(client.clj:296)
    at aleph.http.client$http_connection$fn__18161$fn__18173.invoke(client.clj:295)
    at aleph.http.client_middleware$wrap_request_timing$fn__18086.invoke(client_middleware.clj:425)
    at aleph.http.client_middleware$wrap_query_params$fn__18031.invoke(client_middleware.clj:305)
    at aleph.http.client_middleware$wrap_basic_auth$fn__18038.invoke(client_middleware.clj:325)
    at aleph.http.client_middleware$wrap_oauth$fn__18042.invoke(client_middleware.clj:335)
    at aleph.http.client_middleware$wrap_user_info$fn__18047.invoke(client_middleware.clj:347)
    at aleph.http.client_middleware$wrap_url$fn__18078.invoke(client_middleware.clj:403)
    at aleph.http.client_middleware$wrap_redirects$fn__17954$fn__17973.invoke(client_middleware.clj:173)
    at manifold.deferred$chain_.invoke(deferred.clj:635)
    at manifold.deferred$chain.invoke(deferred.clj:704)
    at aleph.http.client_middleware$wrap_redirects$fn__17954.invoke(client_middleware.clj:173)
    at aleph.http.client_middleware$wrap_exceptions$fn__17911$fn__17921.invoke(client_middleware.clj:123)
    at manifold.deferred$chain_.invoke(deferred.clj:635)
    at manifold.deferred$chain.invoke(deferred.clj:704)
    at aleph.http.client_middleware$wrap_exceptions$fn__17911.invoke(client_middleware.clj:123)
    at aleph.http.client_middleware$wrap_accept$fn__18006.invoke(client_middleware.clj:249)
    at aleph.http.client_middleware$wrap_accept_encoding$fn__18012.invoke(client_middleware.clj:263)
    at aleph.http.client_middleware$wrap_content_type$fn__18001.invoke(client_middleware.clj:239)
    at aleph.http.client_middleware$wrap_form_params$fn__18057.invoke(client_middleware.clj:369)
    at aleph.http.client_middleware$wrap_nested_params$fn__18074.invoke(client_middleware.clj:393)
    at aleph.http.client_middleware$wrap_method$fn__18052.invoke(client_middleware.clj:356)
    at aleph.http.client_middleware$wrap_unknown_host$fn__18082.invoke(client_middleware.clj:412)
    at aleph.http.client_middleware$wrap_request$fn__18093.invoke(client_middleware.clj:459)
    at aleph.http$request$fn__18252$fn__18255$fn__18256.invoke(http.clj:204)
    at manifold.deferred$chain_.invoke(deferred.clj:635)
    at manifold.deferred$chain.invoke(deferred.clj:704)
    at aleph.http$request$fn__18252$fn__18255.invoke(http.clj:204)
    at manifold.deferred$chain_.invoke(deferred.clj:635)
    at manifold.deferred$chain_$fn__13186.invoke(deferred.clj:631)
    at manifold.deferred.Listener$f__12755__auto____12960.invoke(deferred.clj:158)
    at manifold.deferred.Listener.onSuccess(deferred.clj:158)
    at manifold.deferred.Deferred$fn__13028.invoke(deferred.clj:322)
    at manifold.deferred.Deferred.success(deferred.clj:322)
    at manifold.deferred$success_BANG_.invoke(deferred.clj:182)
    at manifold.deferred$catch$fn__13223.invoke(deferred.clj:742)
    at manifold.deferred.Listener$f__12755__auto____12960.invoke(deferred.clj:158)
    at manifold.deferred.Listener.onSuccess(deferred.clj:158)
    at manifold.deferred.Deferred$fn__13028.invoke(deferred.clj:322)
    at manifold.deferred.Deferred.success(deferred.clj:322)
    at manifold.deferred$success_BANG_.invoke(deferred.clj:182)
    at manifold.deferred$chain_.invoke(deferred.clj:643)
    at manifold.deferred$chain_$fn__13186.invoke(deferred.clj:631)
    at manifold.deferred.Listener$f__12755__auto____12960.invoke(deferred.clj:158)
    at manifold.deferred.Listener.onSuccess(deferred.clj:158)
    at manifold.deferred.Deferred$fn__13028.invoke(deferred.clj:322)
    at manifold.deferred.Deferred.success(deferred.clj:322)
    at manifold.deferred$success_BANG_.invoke(deferred.clj:182)
    at manifold.deferred$chain_.invoke(deferred.clj:643)
    at manifold.deferred$chain_$fn__13186.invoke(deferred.clj:631)
    at manifold.deferred.Listener$f__12755__auto____12960.invoke(deferred.clj:158)
    at manifold.deferred.Listener.onSuccess(deferred.clj:158)
    at manifold.deferred.Deferred$fn__13028.invoke(deferred.clj:322)
    at manifold.deferred.Deferred.success(deferred.clj:322)
    at manifold.deferred$success_BANG_.invoke(deferred.clj:182)
    at manifold.deferred$chain_.invoke(deferred.clj:643)
    at manifold.deferred$chain_$fn__13186.invoke(deferred.clj:631)
    at manifold.deferred.Listener$f__12755__auto____12960.invoke(deferred.clj:158)
    at manifold.deferred.Listener.onSuccess(deferred.clj:158)
    at manifold.deferred.Deferred$fn__13028.invoke(deferred.clj:322)
    at manifold.deferred.Deferred.success(deferred.clj:322)
    at manifold.deferred$success_BANG_.invoke(deferred.clj:182)
    at manifold.deferred$chain_.invoke(deferred.clj:643)
    at manifold.deferred$chain_$fn__13186.invoke(deferred.clj:631)
    at manifold.deferred.Listener$f__12755__auto____12960.invoke(deferred.clj:158)
    at manifold.deferred.Listener.onSuccess(deferred.clj:158)
    at manifold.deferred.Deferred$fn__13028.invoke(deferred.clj:322)
    at manifold.deferred.Deferred.success(deferred.clj:322)
    at manifold.deferred$success_BANG_.invoke(deferred.clj:182)
    at aleph.netty$wrap_future$reify__16634.operationComplete(netty.clj:158)
    at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:680)
    at io.netty.util.concurrent.DefaultPromise.notifyListeners0(DefaultPromise.java:603)
    at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:563)
    at io.netty.util.concurrent.DefaultPromise.trySuccess(DefaultPromise.java:406)
    at io.netty.channel.DefaultChannelPromise.trySuccess(DefaultChannelPromise.java:82)
    at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.fulfillConnectPromise(AbstractNioChannel.java:247)
    at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:282)
    at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:528)
    at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:468)
    at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:382)
    at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:354)
    at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:116)
    at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:137)
    at java.lang.Thread.run(Thread.java:745)

Apr 14, 2015 10:37:55 AM manifold.stream.graph invoke
SEVERE: error in message propagation
java.lang.NullPointerException
    at clojure.core$name.invoke(core.clj:1522)
    at aleph.http.core$ring_request__GT_netty_request.invoke(core.clj:157)
    at aleph.http.client$http_connection$fn__18161$fn__18162.invoke(client.clj:278)
    at manifold.stream.Callback.put(stream.clj:431)
    at manifold.stream.graph$async_send.invoke(graph.clj:50)
    at manifold.stream.graph$async_connect$this__13640.invoke(graph.clj:172)
    at manifold.stream.graph$async_connect$this__13640$fn__13641$fn__13642.invoke(graph.clj:153)
    at clojure.core$trampoline.invoke(core.clj:6033)
    at manifold.stream.graph$async_connect$this__13640$fn__13641.invoke(graph.clj:153)
    at manifold.deferred.Listener$f__12755__auto____12960.invoke(deferred.clj:158)
    at manifold.deferred.Listener.onSuccess(deferred.clj:158)
    at manifold.deferred.Deferred$fn__13037.invoke(deferred.clj:324)
    at manifold.deferred.Deferred.success(deferred.clj:324)
    at manifold.deferred$success_BANG_.invoke(deferred.clj:184)
    at manifold.stream.default.Stream$fn__13733.invoke(default.clj:96)
    at manifold.stream.default.Stream.put(default.clj:95)
    at manifold.stream.default.Stream.put(default.clj:119)
    at aleph.http.client$http_connection$fn__18161$fn__18173$fn__18174.invoke(client.clj:296)
    at aleph.http.client$http_connection$fn__18161$fn__18173.invoke(client.clj:295)
    at aleph.http.client_middleware$wrap_request_timing$fn__18086.invoke(client_middleware.clj:425)
    at aleph.http.client_middleware$wrap_query_params$fn__18031.invoke(client_middleware.clj:305)
    at aleph.http.client_middleware$wrap_basic_auth$fn__18038.invoke(client_middleware.clj:325)
    at aleph.http.client_middleware$wrap_oauth$fn__18042.invoke(client_middleware.clj:335)
    at aleph.http.client_middleware$wrap_user_info$fn__18047.invoke(client_middleware.clj:347)
    at aleph.http.client_middleware$wrap_url$fn__18078.invoke(client_middleware.clj:403)
    at aleph.http.client_middleware$wrap_redirects$fn__17954$fn__17973.invoke(client_middleware.clj:173)
    at manifold.deferred$chain_.invoke(deferred.clj:635)
    at manifold.deferred$chain.invoke(deferred.clj:704)
    at aleph.http.client_middleware$wrap_redirects$fn__17954.invoke(client_middleware.clj:173)
    at aleph.http.client_middleware$wrap_exceptions$fn__17911$fn__17921.invoke(client_middleware.clj:123)
    at manifold.deferred$chain_.invoke(deferred.clj:635)
    at manifold.deferred$chain.invoke(deferred.clj:704)
    at aleph.http.client_middleware$wrap_exceptions$fn__17911.invoke(client_middleware.clj:123)
    at aleph.http.client_middleware$wrap_accept$fn__18006.invoke(client_middleware.clj:249)
    at aleph.http.client_middleware$wrap_accept_encoding$fn__18012.invoke(client_middleware.clj:263)
    at aleph.http.client_middleware$wrap_content_type$fn__18001.invoke(client_middleware.clj:239)
    at aleph.http.client_middleware$wrap_form_params$fn__18057.invoke(client_middleware.clj:369)
    at aleph.http.client_middleware$wrap_nested_params$fn__18074.invoke(client_middleware.clj:393)
    at aleph.http.client_middleware$wrap_method$fn__18052.invoke(client_middleware.clj:356)
    at aleph.http.client_middleware$wrap_unknown_host$fn__18082.invoke(client_middleware.clj:412)
    at aleph.http.client_middleware$wrap_request$fn__18093.invoke(client_middleware.clj:459)
    at aleph.http$request$fn__18252$fn__18255$fn__18256.invoke(http.clj:204)
    at manifold.deferred$chain_.invoke(deferred.clj:635)
    at manifold.deferred$chain.invoke(deferred.clj:704)
    at aleph.http$request$fn__18252$fn__18255.invoke(http.clj:204)
    at manifold.deferred$chain_.invoke(deferred.clj:635)
    at manifold.deferred$chain_$fn__13186.invoke(deferred.clj:631)
    at manifold.deferred.Listener$f__12755__auto____12960.invoke(deferred.clj:158)
    at manifold.deferred.Listener.onSuccess(deferred.clj:158)
    at manifold.deferred.Deferred$fn__13028.invoke(deferred.clj:322)
    at manifold.deferred.Deferred.success(deferred.clj:322)
    at manifold.deferred$success_BANG_.invoke(deferred.clj:182)
    at manifold.deferred$catch$fn__13223.invoke(deferred.clj:742)
    at manifold.deferred.Listener$f__12755__auto____12960.invoke(deferred.clj:158)
    at manifold.deferred.Listener.onSuccess(deferred.clj:158)
    at manifold.deferred.Deferred$fn__13028.invoke(deferred.clj:322)
    at manifold.deferred.Deferred.success(deferred.clj:322)
    at manifold.deferred$success_BANG_.invoke(deferred.clj:182)
    at manifold.deferred$chain_.invoke(deferred.clj:643)
    at manifold.deferred$chain_$fn__13186.invoke(deferred.clj:631)
    at manifold.deferred.Listener$f__12755__auto____12960.invoke(deferred.clj:158)
    at manifold.deferred.Listener.onSuccess(deferred.clj:158)
    at manifold.deferred.Deferred$fn__13028.invoke(deferred.clj:322)
    at manifold.deferred.Deferred.success(deferred.clj:322)
    at manifold.deferred$success_BANG_.invoke(deferred.clj:182)
    at manifold.deferred$chain_.invoke(deferred.clj:643)
    at manifold.deferred$chain_$fn__13186.invoke(deferred.clj:631)
    at manifold.deferred.Listener$f__12755__auto____12960.invoke(deferred.clj:158)
    at manifold.deferred.Listener.onSuccess(deferred.clj:158)
    at manifold.deferred.Deferred$fn__13028.invoke(deferred.clj:322)
    at manifold.deferred.Deferred.success(deferred.clj:322)
    at manifold.deferred$success_BANG_.invoke(deferred.clj:182)
    at manifold.deferred$chain_.invoke(deferred.clj:643)
    at manifold.deferred$chain_$fn__13186.invoke(deferred.clj:631)
    at manifold.deferred.Listener$f__12755__auto____12960.invoke(deferred.clj:158)
    at manifold.deferred.Listener.onSuccess(deferred.clj:158)
    at manifold.deferred.Deferred$fn__13028.invoke(deferred.clj:322)
    at manifold.deferred.Deferred.success(deferred.clj:322)
    at manifold.deferred$success_BANG_.invoke(deferred.clj:182)
    at manifold.deferred$chain_.invoke(deferred.clj:643)
    at manifold.deferred$chain_$fn__13186.invoke(deferred.clj:631)
    at manifold.deferred.Listener$f__12755__auto____12960.invoke(deferred.clj:158)
    at manifold.deferred.Listener.onSuccess(deferred.clj:158)
    at manifold.deferred.Deferred$fn__13028.invoke(deferred.clj:322)
    at manifold.deferred.Deferred.success(deferred.clj:322)
    at manifold.deferred$success_BANG_.invoke(deferred.clj:182)
    at aleph.netty$wrap_future$reify__16634.operationComplete(netty.clj:158)
    at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:680)
    at io.netty.util.concurrent.DefaultPromise.notifyListeners0(DefaultPromise.java:603)
    at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:563)
    at io.netty.util.concurrent.DefaultPromise.trySuccess(DefaultPromise.java:406)
    at io.netty.channel.DefaultChannelPromise.trySuccess(DefaultChannelPromise.java:82)
    at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.fulfillConnectPromise(AbstractNioChannel.java:247)
    at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:282)
    at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:528)
    at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:468)
    at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:382)
    at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:354)
    at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:116)
    at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:137)
    at java.lang.Thread.run(Thread.java:745)
```

I'm not sure how to do that without being a performance drag, though. In my particular application, illegal requests should _normally_ be impossible (I do schema verification; but methods simply hadn't been part of my schema so far), so I guess I'd also want the fast-path-I-pinky-swear-my-request-map-is-good version?

**Comment by ztellman** [comment]
All the middleware we have means that one more that checks for keys won't matter, performance-wise.  There should be some sort of `wrap-validation` middleware.

**Comment by arnaudgeiser** [comment]
I just did a test with the latest version of Aleph.
The problem still exists...

```clojure
@(http/request {:url (str "http://localhost:" port "/echo")})
```

```
1. Unhandled java.lang.NullPointerException
   Cannot invoke "clojure.lang.Named.getName()" because "x" is null

                  core.clj: 1597  clojure.core/name
                  core.clj: 1591  clojure.core/name
                  core.clj:  181  aleph.http.core/ring-request->netty-request
                  core.clj:  177  aleph.http.core/ring-request->netty-request
                client.clj:  522  aleph.http.client/http-connection/fn/fn
                stream.clj:  454  manifold.stream.Callback/put
                 graph.clj:   82  manifold.stream.graph/async-send
                 graph.clj:   78  manifold.stream.graph/async-send
                 graph.clj:  208  manifold.stream.graph/async-connect/this
                 graph.clj:  189  manifold.stream.graph/async-connect/this/fn/fn
                  core.clj: 6299  clojure.core/trampoline
                  core.clj: 6288  clojure.core/trampoline
                 graph.clj:  189  manifold.stream.graph/async-connect/this/fn
              deferred.clj:  220  manifold.deferred.Listener/onSuccess
              deferred.clj:  401  manifold.deferred.Deferred/fn
              deferred.clj:  401  manifold.deferred.Deferred/success
              deferred.clj:  246  manifold.deferred/success!
              deferred.clj:  241  manifold.deferred/success!
               default.clj:  173  manifold.stream.default.Stream/fn
               default.clj:  159  manifold.stream.default.Stream/put
               default.clj:  195  manifold.stream.default.Stream/put
                client.clj:  583  aleph.http.client/http-connection/fn/fn/fn
                client.clj:  582  aleph.http.client/http-connection/fn/fn
     client_middleware.clj:  242  aleph.http.client-middleware/wrap-exceptions/fn/fn
                  AFn.java:  154  clojure.lang.AFn/applyToHelper
                  AFn.java:  144  clojure.lang.AFn/applyTo
                  core.clj:  667  clojure.core/apply
                  core.clj: 1977  clojure.core/with-bindings*
                  core.clj: 1977  clojure.core/with-bindings*
               RestFn.java:  142  clojure.lang.RestFn/applyTo
                  core.clj:  671  clojure.core/apply
                  core.clj: 2007  clojure.core/bound-fn*/fn
               RestFn.java:  408  clojure.lang.RestFn/invoke
              deferred.clj:  750  manifold.deferred/eval9611/chain'-
     client_middleware.clj:  242  aleph.http.client-middleware/wrap-exceptions/fn
     client_middleware.clj:  577  aleph.http.client-middleware/wrap-request-timing/fn
     client_middleware.clj:  950  aleph.http.client-middleware/wrap-request/fn
                  http.clj:  348  aleph.http/eval15602/request/fn/fn/fn
              deferred.clj:  768  manifold.deferred/eval9611/chain'-
              deferred.clj:  722  manifold.deferred/eval9611/subscribe/fn
              deferred.clj:  220  manifold.deferred.Listener/onSuccess
              deferred.clj:  399  manifold.deferred.Deferred/fn/fn
                  AFn.java:   22  clojure.lang.AFn/run
             Executor.java:  318  io.aleph.dirigiste.Executor$3/run
             Executor.java:   62  io.aleph.dirigiste.Executor$Worker$1/run
              executor.clj:   70  manifold.executor/thread-factory/reify/f
                  AFn.java:   22  clojure.lang.AFn/run
               Thread.java:  833  java.lang.Thread/run
```

**Comment by arnaudgeiser** [comment]
Considering `DefaultHttpRequest` [1] requires an `HttpMethod` to be present, it can be even simpler, we just have to throw an Exception that is explicit regarding the missing parameter.

[1] : https://netty.io/4.1/api/io/netty/handler/codec/http/DefaultHttpRequest.html

**Comment by arnaudgeiser** [comment]
Closed by #679
